sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ArrayExpressionUtils.java

```diff
@@ -19,158 +19,288 @@
+  private static final Comparator<Byte> byteComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    byte c1 = (Byte) o1, c2 = (Byte) o2;
-    return Byte.compare(c1, c2);
+    return Byte.compare(o1, o2);
   };
 
-  private static final Comparator<Object> shortComp = (o1, o2) -> {
+  // Short ascending nullable comparator
+  private static final Comparator<Short> shortComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    short c1 = (Short) o1, c2 = (Short) o2;
-    return Short.compare(c1, c2);
+    return Short.compare(o1, o2);
   };
 
-  private static final Comparator<Object> integerComp = (o1, o2) -> {
+  // Integer ascending nullable comparator
+  private static final Comparator<Integer> integerComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    int c1 = (Integer) o1, c2 = (Integer) o2;
-    return Integer.compare(c1, c2);
+    return Integer.compare(o1, o2);
   };
 
-  private static final Comparator<Object> longComp = (o1, o2) -> {
+  // Long ascending nullable comparator
+  private static final Comparator<Long> longComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    long c1 = (Long) o1, c2 = (Long) o2;
-    return Long.compare(c1, c2);
+    return Long.compare(o1, o2);
   };
 
-  private static final Comparator<Object> floatComp = (o1, o2) -> {
+  // Float ascending nullable comparator
+  private static final Comparator<Float> floatComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    float c1 = (Float) o1, c2 = (Float) o2;
-    return SQLOrderingUtil.compareFloats(c1, c2);
+    return SQLOrderingUtil.compareFloats(o1, o2);
   };
 
-  private static final Comparator<Object> doubleComp = (o1, o2) -> {
+  // Double ascending nullable comparator
+  private static final Comparator<Double> doubleComp = (o1, o2) -> {
     if (o1 == null && o2 == null) {
       return 0;
     } else if (o1 == null) {
       return -1;
     } else if (o2 == null) {
       return 1;
     }
-    double c1 = (Double) o1, c2 = (Double) o2;
-    return SQLOrderingUtil.compareDoubles(c1, c2);
+    return SQLOrderingUtil.compareDoubles(o1, o2);
   };
 
-  public static int binarySearchNullSafe(ArrayData data, Boolean value) {
-    return Arrays.binarySearch(data.toObjectArray(BooleanType$.MODULE$), value, booleanComp);
+  // boolean
+  // boolean non-nullable
+  public static int binarySearch(boolean[] data, boolean value) {
```

Comment:
since the array is sorted, I think it should be pretty easy to find a boolean value from it, by just checking the first or the last element.

Reply:
@cloud-fan 
Do we need to consider the following scenarios?
<img width="633" alt="image" src="https://github.com/user-attachments/assets/73e6348b-9aea-4831-8132-d30808501634">

Because like `Arrays.binarySearch(int[] a, int key)` also support `multiple elements` in order.
<img width="621" alt="image" src="https://github.com/user-attachments/assets/fe5e19b3-2a63-41a8-8326-d63f00b22d60">

Reply:
The above example (with `duplicate elements`) should return `different result values` when using `check head and tail` and `standard binary search`, although ` If the array contains multiple elements with the specified value, there is no guarantee which one will be found.`

Reply:
I see, we should have consistent result between the primitive and boxed versions

Reply:
Yea! ❤️